test/unittests/Connection_test.cpp

```diff
@@ -0,0 +1,134 @@
+/**
+ * (C) 2017 KISTLER INSTRUMENTE AG, Winterthur, Switzerland
+ *
+ * @file Connection_test.cpp
+ * @author Ardazishvili Roman (ardazishvili.roman@yandex.ru)
+ *
+ * Created on: Feb 4, 2019
+ * Project: sdbus-c++
+ * Description: High-level D-Bus IPC C++ library based on sd-bus
+ *
+ * This file is part of sdbus-c++.
+ *
+ * sdbus-c++ is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * sdbus-c++ is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with sdbus-c++. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <gtest/gtest.h>
+
+#include "Connection.h"
+#include "mocks/Dbus_interface_mock.h"
+
+using ::testing::_;
+using ::testing::DoAll;
+using ::testing::SetArgPointee;
+using ::testing::Return;
+
+using BusType = sdbus::internal::Connection::BusType;
+
+
+class ConnectionCreationTest : public ::testing::Test {
+protected:
+    ConnectionCreationTest() = default;
+
+    std::unique_ptr<Dbus_interface_mock> mock_ { std::make_unique<Dbus_interface_mock>(Dbus_interface_mock()) };
+    sd_bus* STUB_ { reinterpret_cast<sd_bus*>(1) };
+};
+
+TEST_F(ConnectionCreationTest, DefaultConstructorNoThrowSystem)
```

Comment:
With such single-purpose tests, we then typically use an assertion or expectation only for that single, specific tested behavior. For other unimportant interactions, we use ON_CALLs or even nothing. For example, concerning OpensAndFlushesBusWhenCreated test:
```
TEST_F(ASystemBusConnection, OpensAndFlushesBusWhenCreated)
{
    EXPECT_CALL(*mock_, sd_bus_open_system(_)).WillOnce(DoAll(SetArgPointee<0>(STUB_), Return(1)));
    EXPECT_CALL(*mock_, sd_bus_flush(_));

    sdbus::internal::Connection{BusType::eSystem, std::move(mock_)};
}
````

Note these differences:
* We only apply expectations on calls that relate to the tested behavior in that test. `sd_bus_flush_close_unref` is not of an interest here (it is in other test - ClosesAndUnrefsBusWhenDestructed), so we don't clutter the test with an unnecessary expectation here, and keep the test at a proper abstraction level.
* We don't need ASSERT_NO_THROW, since non-throwing IMHO is a default, expected happy path behavior. If not, then we shall create another test just for testing non-throwing for happy path, e.g. DoesNotThrowWhenCreatedSuccessfully. But that's an overkill in a good design using exceptions only for reporting exceptional errors.
* Using EXPECT_CALL a lot, with lots of constraints, leads to [over-specified, brittle tests](https://github.com/google/googletest/blob/master/googlemock/docs/CookBook.md#knowing-when-to-expect), that break that are tied to code implementation, not its contract. When we change implementation, the test breaks, and that's undesirable.

So, in general, it's important not only to not have shallow tests, but also not too detailed, implementation-dependent tests. Balance is the key here.